Sources/AblyAssetTrackingInternal/AblyWrapper/DefaultAbly.swift

```diff
@@ -0,0 +1,325 @@
+import CoreLocation
+import Ably
+import AblyAssetTrackingCore
+import Logging
+
+public class DefaultAbly: AblyCommon {
+    
+    public weak var publisherDelegate: AblyPublisherServiceDelegate?
+    public weak var subscriberDelegate: AblySubscriberServiceDelegate?
+    
+    private let logger: Logger
+    private let client: ARTRealtime
+    private let mode: AblyMode
+    
+    private var channels: [String: ARTRealtimeChannel] = [:]
+
+    public required init(configuration: ConnectionConfiguration, mode: AblyMode, logger: Logger) {
+        self.client = ARTRealtime(options: configuration.getClientOptions())
+        self.mode = mode
+        self.logger = logger
+    }
+    
+    public func connect(
+        trackableId: String,
+        presenceData: PresenceData,
+        useRewind: Bool,
+        completion: @escaping ResultHandler<Void>
+    ) {
+        guard channels[trackableId] == nil else {
+            completion(.success)
+            
+            return
+        }
+        
+        let options = ARTRealtimeChannelOptions()
+        options.modes = [.presenceSubscribe, .presence]
+        
+        if useRewind {
+            options.params = ["rewind": "1"]
+        }
+        
+        if mode.contains(.subscribe) {
+            options.modes.insert(.subscribe)
+        }
+        
+        if mode.contains(.publish) {
+            options.modes.insert(.publish)
+        }
+        
+        let channel = client.channels.getChannelFor(trackingId: trackableId, options: options)
+        
+        // Force cast intentional here. It's a fatal error if we are unable to create presenceData JSON
+        channel.presence.enter(try! presenceData.toJSONString()) { [weak self] error in
+            guard let error = error else {
+                self?.logger.debug("Entered to channel [id: \(trackableId)] presence successfully", source: String(describing: Self.self))
+                self?.channels[trackableId] = channel
+                completion(.success)
+                
+                return
+            }
+            
+            self?.logger.error("Error during joining to channel [id: \(trackableId)] presence: \(String(describing: error))", source: String(describing: Self.self))
+            completion(.failure(error.toErrorInformation()))
+        }
+    }
+    
+    public func disconnect(trackableId: String, presenceData: PresenceData, completion: @escaping ResultHandler<Bool>) {
+        guard let channelToRemove = channels[trackableId] else {
+            completion(.success(false))
+            
+            return
+        }
+        
+        // Force cast intentional here. It's a fatal error if we are unable to create presenceData JSON
+        //
+        channelToRemove.presence.leave(try! presenceData.toJSONString()) { [weak self] error in
```

Comment:
There are not gonna be any useful crash message here, just "nil on force unwrap". Maybe explicitly add `precondition` crash in these cases? @lawrence-forooghian  WDYT

Reply:
Agree, I'll change it

Reply:
1fa230ac04856cfa97143ea13e3158d5961b825c